File: codes/Garnetwzy/64.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        if(m == 0)
            return 0;
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[0][0] = grid[0][0];
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(i == 0 && j == 0)
                    continue;
                int first = INT_MAX;
                int second = INT_MAX;
                if(i > 0) {
                    first = dp[i-1][j];
                }
                if(j > 0) {
                    second = dp[i][j-1];
                }
                dp[i][j] = min(first, second) + grid[i][j];
            }
        }
        return dp[m-1][n-1];
    }
};

File: codes/HellDream/64.java
class Solution {
    public int minPathSum(int[][] grid) {
        int rows = grid.length;
        int cols=  grid[0].length;
        int[][] costs = new int[rows][cols];
        costs[0][0] = grid[0][0];
        for(int i=1;i<rows;i++){
            costs[i][0] = costs[i-1][0]+grid[i][0];
        }
        for(int i=1;i<cols;i++){
            costs[0][i] = costs[0][i-1]+grid[0][i];
        }
        for(int row=1;row<rows;row++){
            for(int col=1;col<cols;col++){
                costs[row][col] = Math.min(costs[row-1][col],costs[row][col-1])+grid[row][col];
            }
        }
        return costs[rows-1][cols-1];
    }
}